### PR TITLE
ADD alternatives validation functionality using config option

### PR DESCRIPTION
**Pull request by lloria91**
1) added **_alternate_** option to config (query, params, headers, body values will be checked)
2) added tests for coverage
3) implementation:
    - checking for query, params, headers, body inside alternate, and constructing new schema and data for Joi.alternatives().try(...). 
    - Validating one by one this data: if any of body, headers, params, query is ok -> alternative validation succeed. If all fail -> alternative fails

### CODE REVIEW

**Comment by fkanout** [comment]
@lloria91 could you please update the readme file with the new changes?

File: src/index.js [hunk 1]
```diff
@@ -12,59 +32,93 @@ const handleErrorWithSource = async (source, fn) => {
...
     denyUnknown: [],
     httpErrorCode: 400,
     nextOnError: false,
+    alternate: [],
     ...config
   };
 
   return async (ctx, next) => {
+    let alternateSchema = {};
+    let alternateData = [];
```
Comment (fkanout) [inline src/index.js:49]: @lloria91 this variable is initiated as an array, but it seems that its value is an object?

Comment (lloria91) [reply src/index.js:49]: yes. after declaration it was changed to an object, so missed it ;)

File: README.md [hunk 2]
```diff
@@ -103,6 +103,18 @@ import validate, { Joi } from ('koa-router-joi-validation');
 
   **default** `false`
 
+#
+
+- `alternate`
```
Comment (fkanout) [inline README.md:108]: @lloria91 if we could put the link of the Joi alternative function, to explain that this is a wrapper of that function

File: test/index.test.js [hunk 3]
```diff
@@ -465,6 +465,124 @@ describe("koa-router-joi-validation", function() {
...
+
+      app.use(bodyParser());
+      app.use(router.routes());
+      server = await new Promise((resolve, reject) => {
+        _server = http.createServer(app.callback());
+        _server.listen(3001, () => resolve(_server));
+      });
+    });
+
+    it("should fail when neither body and query params are passed correctly", async () => {
```
Comment (fkanout) [inline test/index.test.js:530]: @lloria91 The test should cover one case at the time

File: test/index.test.js [hunk 4]
```diff
@@ -465,6 +465,124 @@ describe("koa-router-joi-validation", function() {
...
+          headers: { "x-unknown-header": true, accept: "application/json", "content-type": "application/json" }
+        });
+      } catch (error) {
+
+        assert.deepEqual(error.response.status, 400);
+        assert.deepEqual(error.response.data, '"query.q" is required');
+      }
+    });
+
+    it("should succed when one of query or body is correct", async () => {
```
Comment (fkanout) [inline test/index.test.js:545]: @lloria91 The test should cover one case at the time

**Comment by fkanout** [comment]
@lloria91 and please check the Lint

File: test/index.test.js (hunk 3 above)
Comment (lloria91) [reply test/index.test.js:530]: @fkanout what do you mean once case at time? Both should fail to fail alternative validation.

File: test/index.test.js (hunk 4 above)
Comment (lloria91) [reply test/index.test.js:545]: @fkanout same here. Should I add test where another param fails and second succeed?

File: src/index.js [hunk 5]
```diff
@@ -1,6 +1,40 @@
...
+          .try({
+            ...schema
+          })
+          .validateAsync(values[v], {
+            allowUnknown: !config.denyUnknown.includes(v)
+          })
+      );
+      return; // returning after first validation succeed
+    } catch (err) {
+      errorMsg += `${err.message}; `;
```
Comment (fkanout) [inline src/index.js:23]: why we need to concatenate the error message? could this be a misleading error message at the end?

Comment (lloria91) [reply src/index.js:23]: I'm giving information about all failed errors. For example why body failed and why query failed

File: src/index.js [hunk 6]
```diff
@@ -12,54 +46,104 @@ const handleErrorWithSource = async (source, fn) => {
...
     nextOnError: false,
+    alternate: [],
     ...config
   };
 
   return async (ctx, next) => {
+    let alternateSchema = {};
+    let alternateData = {};
+
+    if (_config.alternate.includes("query")) {
```
[the remark this answers is not in the document]
Comment (lloria91) [reply src/index.js:68]: You are right, dont have tests for it. I will add this check for alternates and denyUnknown too

File: src/index.js (hunk 5 above)
Comment (fkanout) [reply src/index.js:23]: got it.

**Pull request merged**